### PR TITLE
packing refactor to use more STL containers

### PR DESCRIPTION
**Pull request by MohamedElgammal**
#### Description

In this PR, I have done some refactoring on the packing code to use more STL containers instead of hand-crafted linked lists. Besides that, I am adding more functions wrappers to t_cluster_placement_stats to prevent direct access to the data members. 

I am also adding a new context to support packing multithreading and some updates to the re-clustering API to support the packing iterative improvement project.

#### Refactored Files
There are multiple files updated in the refactor but the main files are cluster_placement.cpp, vpr_types.cpp, vpr_types.h

### CODE REVIEW

File: vpr/src/base/vpr_types.h [hunk 1]
```diff
@@ -425,14 +425,48 @@ struct t_chain_info {
...
-    t_cluster_placement_primitive* in_flight;         ///<ptrs to primitives currently being considered
-    t_cluster_placement_primitive* tried;             ///<ptrs to primitives that are open but current logic block unable to pack to
-    t_cluster_placement_primitive* invalid;           ///<ptrs to primitives that are invalid
+class t_cluster_placement_stats {
+  public:
+    int num_pb_types;                     ///<num primitive pb_types inside complex block
+    bool has_long_chain;                  ///<specifies if this cluster has a molecule placed in it that belongs to a long chain (a chain that spans more than one cluster)
+    const t_pack_molecule* curr_molecule; ///<current molecule being considered for packing
+
+    ///<[0..num_pb_types-1] ptrs to linked list of valid primitives, for convenience, each linked list head is empty
```
Comment (MohamedElgammal) [inline vpr/src/base/vpr_types.h:434]: update comment

File: vpr/src/base/vpr_types.h [hunk 2]
```diff
@@ -425,14 +425,48 @@ struct t_chain_info {
...
+    std::vector<std::unordered_map<int, t_cluster_placement_primitive*>> valid_primitives;
+
+  public:
+    // Moves primitives that are inflight to the tried map
+    void move_inflight_to_tried();
+
+    // Move the primitive at (it) to inflight and increment the current iterator
+    void move_primitive_to_inflight(int pb_type_index, std::unordered_multimap<int, t_cluster_placement_primitive*>::iterator& it);
```
Comment (MohamedElgammal) [inline vpr/src/base/vpr_types.h:442]: why you increment the iterator

**Comment by vaughnbetz** [comment]
Should add a QoR comparison (VTR, Titan anyway) so we know there are no surprises.

File: vpr/src/pack/cluster_placement.cpp [hunk 3]
```diff
@@ -132,57 +118,53 @@ bool get_next_primitive_list(t_cluster_placement_stats* cluster_placement_stats,
...
-                /* remove invalid nodes lazily when encountered */
-                while (cur && cur->valid == false) {
-                    prev->next_primitive = cur->next_primitive;
-                    cur->next_primitive = cluster_placement_stats->invalid;
-                    cluster_placement_stats->invalid = cur;
-                    cur = prev->next_primitive;
-                }
-                if (cur == nullptr) {
-                    break;
+        if (!cluster_placement_stats->valid_primitives[i].empty() && primitive_type_feasible(molecule->atom_block_ids[molecule->root], cluster_placement_stats->valid_primitives[i].begin()->second->pb_graph_node->pb_type)) {
```
Comment (vaughnbetz) [inline vpr/src/pack/cluster_placement.cpp:128]: Maybe a temporary variable for the long chain?

File: vpr/src/pack/re_cluster_util.cpp [hunk 4]
```diff
@@ -86,14 +93,17 @@ void commit_mol_move(const ClusterBlockId& old_clb,
     }
 }
 
-t_lb_router_data* lb_load_router_data(std::vector<t_lb_type_rr_node>* lb_type_rr_graphs, const ClusterBlockId& clb_index, const std::vector<AtomBlockId>& clb_atoms) {
+t_lb_router_data* lb_load_router_data(std::vector<t_lb_type_rr_node>* lb_type_rr_graphs, const ClusterBlockId& clb_index, const std::unordered_set<AtomBlockId>* clb_atoms) {
```
Comment (vaughnbetz) [inline vpr/src/pack/re_cluster_util.cpp:96]: Update (or add) comment in header file about what clb_atoms being NULL means.

**Review by vaughnbetz — CHANGES_REQUESTED** [review]

File: vpr/src/base/vpr_types.cpp [hunk 5]
```diff
@@ -228,39 +228,95 @@ void free_pack_molecules(t_pack_molecule* list_of_pack_molecules) {
...
-        for (int j = 0; j < cluster_placement_stats_list[index].num_pb_types; j++) {
-            cur = cluster_placement_stats_list[index].valid_primitives[j]->next_primitive;
-            while (cur != nullptr) {
-                next = cur->next_primitive;
-                delete cur;
-                cur = next;
-            }
-            delete cluster_placement_stats_list[index].valid_primitives[j];
+        if (cluster_placement_primitive.second->pb_graph_node->pb_type
+            == valid_primitives[i].begin()->second->pb_graph_node->pb_type) {
```
[the remark this answers is not in the document]
Comment (MohamedElgammal) [reply vpr/src/base/vpr_types.cpp]: I have double-checked and found that t_pb_type doesn't have an index member. Hence, on the current data structure we should be checking that the type is the same. In future refactoring, a unique index can be added to each primitive pb_type (unique over its cluster block type) and this index can be used directly here.

**Comment by mithro** [comment]
@hzeller - If you get a moment, as a C++ expert can you give this pull request a once over?

**Comment by vaughnbetz** [comment]
Mohamed, can you run another seed on the vtr benchmarks to confirm the pack time ratios make sense -- the place times were highly variable so there may be some machine load effects going on.

**Comment by MohamedElgammal** [comment]
@vaughnbetz I have added the run for another seed on a new sheet at the same spreadsheet. It shows that the pack time and place time are both almost equal to that of the master branch before the refactor. To mitigate the machine load effect in this run, I have run the master and the current branch simultaneously.

**Comment by vaughnbetz** [comment]
Looks good, thanks.

**Comment by vaughnbetz** [comment]
A couple of tests didn't run (vtr_reg_nightly3) and there's one yosys + partial mapping QoR change (looks benign). Updating to see if they go away, but this looks safe to merge.

**Pull request merged**